File: src/main/java/com/gusrylmubarok/reddit/backend/controller/SubredditController.java
package com.gusrylmubarok.reddit.backend.controller;

import com.gusrylmubarok.reddit.backend.dto.SubredditDto;
import com.gusrylmubarok.reddit.backend.service.SubredditService;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;

@RestController
@RequestMapping("/api/v1/subreddit")
@AllArgsConstructor
@Slf4j
public class SubredditController {

    private final SubredditService subredditService;

    @PostMapping
    public ResponseEntity<SubredditDto> createSubreddit(@RequestBody @Valid SubredditDto subredditDto)
            throws Exception {
        try {
            return new ResponseEntity<>(subredditService.save(subredditDto), HttpStatus.OK);
        }catch (Exception e) {
            throw new Exception("Error creating subreddit");
        }
    }

    @GetMapping()
    public ResponseEntity<List<SubredditDto>> getAllSubreddits() {
        return new ResponseEntity<>(subredditService.getAll(), HttpStatus.OK);
    }

    @GetMapping("/{id}")
    public ResponseEntity<SubredditDto> getSubreddit(@PathVariable Long id) {
        return new ResponseEntity<>(subredditService.getSubreddit(id), HttpStatus.OK);
    }
}

File: src/main/kotlin/com/reddot/controller/SubredditController.kt
package com.reddot.controller

import com.reddot.common.RestResult
import com.reddot.data.model.SubredditDTO
import com.reddot.service.SubredditService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/subreddit")
class SubredditController(
    private val subredditService: SubredditService
) {

    @GetMapping
    fun getAllSubbreddit(): ResponseEntity<Any> {
        val result: List<SubredditDTO> =  subredditService.getAll()
        return RestResult.build("success", HttpStatus.OK, result)
    }

    @GetMapping("/{id}")
    fun getSubreddit(@PathVariable id: Long): ResponseEntity<Any> {
        val result: SubredditDTO = subredditService.getSubreddit(id)
        return RestResult.build("success", HttpStatus.OK, result)
    }

    @PostMapping
    fun createSubreddit(@RequestBody subredditDTO: SubredditDTO): ResponseEntity<Any> {
        val result: SubredditDTO = subredditService.save(subredditDTO)
        return RestResult.build("success", HttpStatus.OK, result)
    }

}